### ISSUE TITLE
Both Google Chrome and Safari have begun to block GRNsight's iframe

### ISSUE THREAD

**Issue by mihirsamdarshi** [dondi/GRNsight#688]
At first it was just Safari, but now Google Chrome has started blocking the iframe for "attempting to load unsafe scripts." This needs to be fixed fairly soon.
<img width="1160" alt="screen shot 2018-10-11 at 1 11 43 pm" src="https://user-images.githubusercontent.com/5462077/46831077-35026b80-cd57-11e8-96af-3aaa562732fa.png">

**Comment by kdahlquist** [comment]
@kdahlquist doesn't see this with Chrome on her machine (version is up to date, Version 69.0.3497.100 (Official Build) (64-bit)).  Need to check with other people to see if this is happening to them.

**Comment by mihirsamdarshi** [comment]
The console shows the exact source of the problem:
<img width="1057" alt="screen shot 2018-10-25 at 2 31 22 pm" src="https://user-images.githubusercontent.com/5462077/47531540-c519d800-d862-11e8-85e9-58a8ad3c23ee.png">

**Comment by johnllopez616** [comment]
According to [this link](https://love2dev.com/blog/this-page-is-trying-to-load-scripts-from-unauthenticated-sources/), this might require serving the pages using HTTPS

**Comment by mihirsamdarshi** [comment]
There are 2 parts to solving this:
1) adding a package so that Express can serve the iframe over HTTPS
- @mihirsamdarshi should get this working via a self-signed certificate
2) we need a certificate for that

**Comment by dondi** [comment]
It looks like #327 may be a duplicate of this issue.

**Comment by kdahlquist** [comment]
Yes, I'll close #327--it's been duplicated a few times actually.

**Comment by mihirsamdarshi** [comment]
@dondi, I believe that I set up all the requirements for the server to be delivered over https. The webpage at https://localhost:5001 is delivered over HTTPS, however any requests are not delivered via SSL, because apparently http://localhost:5000 is not over HTTP, despite me setting it up so

**Comment by dondi** [comment]
@mihirsamdarshi There are still a lot of specific accesses of plain HTTP in the source code. For example, here: https://github.com/dondi/GRNsight/blob/beta/server/config/config.js#L5-L14

Try transitioning all of those to `https://` and see if that makes a difference.

**Comment by mihirsamdarshi** [comment]
@dondi I figured it out, just waiting on certificates from Masao

**Comment by dondi** [comment]
OK sounds good, I will check with him.

Of note, certificates are domain-specific. So, whatever you did to get things to work on your specific computer will remain true for other developers. The certificate we acquire will pertain _only_ too grnsight.cs.lmu.edu and will require specific installation on that host. This certificate will not be of use for developers running a local version.

Thus, plan on documenting what you did so that future developers will know how to set up their computers for GRNsight development. It will also be a good idea to look ahead and also document how to set up a permanent server like grnsight.cs.lmu.edu. (I presume that it will be very similar to local computer setup, but just using a different certificate)

**Comment by kdahlquist** [comment]
How about creating a wiki page (or adding to one of the ones we have)...

**Comment by mihirsamdarshi** [comment]
Will do am mocking it up in the setting up GRNsight wiki page right now @dondi @kdahlquist

**Comment by mihirsamdarshi** [comment]
[Both have been added to the wiki](https://github.com/dondi/GRNsight/wiki/Server-Setup), although I am not sure yet about the protocol for the production and beta servers yet.

**Comment by dondi** [comment]
The write-ups generally look right, though I have one question: Does every developer need their own self-signed .key and .cert files, or can one fixed set be created then committed, for all developers to use? The files are known to be for developer use only anyway, so I don't have a problem with putting them on the repository since they'll only work with local machine setups.

That is, presuming that it works. See what you can find out in that department.

**Comment by mihirsamdarshi** [comment]
@dondi it seems that each developer needs their own self-signed certificate, as it seems that there are security concerns with pre-shared self-signed certificates, although I'm not sure what those concerns are

**Comment by dondi** [comment]
Masao has gotten back to @mihirsamdarshi and me with the certificate info and upon looking at what he did, it appears that he was able to do this at a level above our actual software; i.e., the grnsight.cs.lmu.edu host is _already_ serving HTTPS _without_ any software-level changes (with respect to security):

<img width="1050" alt="image" src="https://user-images.githubusercontent.com/382242/58854209-767eb880-8651-11e9-9080-5d7fc513f119.png">

The benefit of this is that it looks like we won’t need to do any further server configuration when deploying GRNsight. Masao has done the work.

The downside is that with the new HTTPS-at-the-level-of-grnsight.cs.lmu.edu setup already having been done, both our release and beta pages are broken (also visible in the screenshot above):

http://dondi.github.io/GRNsight/
http://dondi.github.io/GRNsight/beta.html

The reason for the breakage is the same (attempt to access HTTP from an HTTPS resource), but the source of the HTTP has propagated to a different place. (i.e., whereas the original error was due to grnsight.cs.lmu.edu being HTTP, this time, that site is now HTTPS and _it_ is complaining about resources being accessed via HTTP)

@mihirsamdarshi The task now shifts to looking at what the above sites are complaining about then tracking down the places where those requests are being made. Most of the time, the change will be a matter of shifting URLs from `http://` to `https://` (or possibly just `//` as mentioned earlier).

There may be some instances where it won’t be that simple. Alert me if you encounter those.

As a starter, attached is the error log that I’m currently getting for http://dondi.github.io/GRNsight/beta.html —I would track down each of those errors to the source of the erroneous request in our code, then try to change that to `//` first, and if that doesn’t work, change to to an explicit `https://`.

Hope these instructions make sense. The trial I’m on isn’t over so I will only be accessible during recesses and breaks.

<img width="1050" alt="image" src="https://user-images.githubusercontent.com/382242/58854320-c8274300-8651-11e9-94f2-d03e650876b3.png">

Because the sites are broken _now_, there may be a need for a “hot fix” where we deploy _just HTTPS-related_ changes. Either that or @mihirsamdarshi you can ask Masao to temporarily back out these server changes until such time that the GRNsight software itself is ready for full HTTPS. The final course of action here will, I think, depend on what you find regarding the error messages that are showing up right now.

**Comment by kdahlquist** [comment]
How about putting an "under construction" type message at the top of the GRNsight master and beta pages explaining that the sites are unavailable due to conversion to https?  It seems that would be preferable than having Masao turn it off again.  Assuming that it will be fixed in the next couple of days.

**Comment by mihirsamdarshi** [comment]
@kdahlquist I did that

@dondi does this mean I should remove the Express changes from my PR?

**Comment by dondi** [comment]
@mihirsamdarshi I’m not sure what will stay or go. The problem with removing the _https_ server for local development is that, if we change all other links to _https_, then I don’t know if a local server instance will work if it can’t do _https_.

Then again, I don’t think having a local Express HTTPS server will work well with the server-level certificate and HTTPS that Masao has set up.

My guess as to what will happen: (a) yes we will no longer need the Express _https_ server, but (b) we will need a configuration change where, when running locally, we use _http_ links but running in production or beta uses _https_ links.

This would be exactly the situation where just plain `//` URLs (i.e., without either `http:` nor `https:` preceding) would be useful, if they worked. There might be some trial and error needed (sorry, no pun intended regarding my current situation).

I think you’ll have to do the following:
* Test a local server to make sure it works without errors
* Upload that to _beta_ to see if that works (if we’re under construction, then having up-and-down functionality will be understandable)
* Rinse and repeat until you have the right combination of software, configuration, and links that will work correctly in both situations

**Comment by dondi** [comment]
Another possibility: set the server up so that it responds to _both_ http and https. That way, the https “wrapper” takes over on the server for production and beta while talking to that server in http, but for local development, we access the https on the Express server directly.

There is some open-endedness here (i.e., I don’t know the answer offhand myself, and would be doing the exact same thing I’m suggesting in these comments if I were tasked directly with resolving this issue). The key is to be aware of what we want to accomplish and to systematically work our way to that situation.

**Comment by mihirsamdarshi** [comment]
understood, thanks

**Comment by mihirsamdarshi** [comment]
I just added the fixes for jQuery not delivering over HTTPS, however I still get the error `Mixed Content: The page at 'https://dondi.github.io/GRNsight/beta.html' was loaded over HTTPS, but requested an insecure XMLHttpRequest endpoint 'http://grnsight.cs.lmu.edu/server/ga?path=beta.html'. This request has been blocked; the content must be served over HTTPS.` when loading the page, meaning that the iframe is still not being delivered over HTTPS. I think that the server level changes, creating a certificate must be done.

**Comment by mihirsamdarshi** [comment]
this is fixed, and the iframe now delivers over HTTPS

**Comment by dondi** [comment]
When the https transition is complete, how should we handle master/production? We can either wait until we have a release, or we “peel off” just the https changes into an interim hotfix on the master branch.

**Comment by mihirsamdarshi** [comment]
Actually, all it took was changing gh-pages, so Master is already fixed as well

**Comment by mihirsamdarshi** [comment]
I have also configured it so that localhost does not need to be served over HTTPS, thus eliminating a need for developers to install their own certificates.

**Comment by dondi** [comment]
I’m still seeing errors in both production and beta; looks the same for production as yesterday, but down to just the `/server/ga` one in beta (see screenshots). (I disabled to cache to make sure I got the latest files)

It looks like the production files haven’t been updated yet? Let me know when the HTTPS transition is complete so we can close this issue. Thanks!

* Production:

<img width="1156" alt="image" src="https://user-images.githubusercontent.com/382242/58984712-8a7c0480-878e-11e9-8b11-bd8e205c5642.png">

* Beta:

<img width="1156" alt="image" src="https://user-images.githubusercontent.com/382242/58984698-80f29c80-878e-11e9-87fb-ce681a1d0661.png">

**Comment by kdahlquist** [comment]
I'm going to leave this one open--on my machine, beta is functional in Firefox and Chrome, but master is not.  I'll let @dondi chime in as to when it is appropriate to close.

**Comment by dondi** [comment]
I’m seeing the same thing as @kdahlquist: production isn’t loading right but beta is. I’m currently not in a situation where I can check out the console log so can’t provide further details for now, but @mihirsamdarshi please verify whether master/production actually needs a re-release. You had mentioned in https://github.com/dondi/GRNsight/issues/688#issuecomment-499190434 that it was not needed, but based on what we’re seeing this might not be the case.

**Comment by kdahlquist** [comment]
@mihirsamdarshi told me that master won't be updated until we do a release.

**Comment by dondi** [comment]
OK, so perhaps we can leave the under construction sign up for Home until then? @mihirsamdarshi let me know which open or new issues need my intervention.

**Comment by mihirsamdarshi** [comment]
@dondi, hopefully this will be gone after the new release, because it is simply that jQuery and Bootstrap are not being delivered over HTTPS, which has already been fixed in beta

**Comment by dondi** [comment]
This does appear to be the case, so let’s note this as something to monitor whenever the release to production has taken place.

**Comment by dondi** [comment]
Possible regression noted in beta 3.1.24:

https://github.com/dondi/GRNsight/issues/406#issuecomment-500032371

**Comment by mihirsamdarshi** [comment]
@dondi not sure what is going on since that error does not occur when hosted locally

**Comment by mihirsamdarshi** [comment]
Never mind, I figured it out, it is Google Analytics code that triggers that request, however I am not sure how to edit it since it isn't hosted in the GRNsight folder.

**Comment by dondi** [comment]
Try changing this one to `https://` or just plain `//`:

https://github.com/dondi/GRNsight/blob/gh-pages/assets/js/ga-report.js#L3

**Comment by mihirsamdarshi** [comment]
I didn't realize `ga-report` was hosted in gh-pages, it is edited and fixed.

**Comment by dondi** [comment]
Neither did I, until I started to look for it.

Always take into account the entire website when triaging requests that have issues, and not just the specific web app code that you’re working with.

**Comment by dondi** [comment]
You need one more change; now that the requesting URL is https, then the CORS header should also be changed to accept https:

https://github.com/dondi/GRNsight/blob/master/server/controllers/ga-controller.js#L46

Change the `http://` on that line to `https://`

**Comment by dondi** [comment]
(not sure if you weren’t seeing that CORS error due to the presence of a CORS-disabling extension that’s installed in the browser; if you do have such an extension, I recommend generally keeping it turned off unless there is really no other way to get a CORS issue to work—otherwise, you won’t see errors that most other users will see)

**Comment by kdahlquist** [comment]
production is fixed upon release of 4.0.0
